internal/scheduler/ingester.go

```diff
@@ -0,0 +1,777 @@
+type Ingester struct {
+	// Used to setup a Pulsar consumer.
+	PulsarClient    pulsar.Client
+	ConsumerOptions pulsar.ConsumerOptions
+	// Write to postgres at least this often (assuming there are records to write).
+	MaxWriteInterval time.Duration
+	// Write current batch to postgres if at least this many records have been written to it.
+	MaxWriteRecords int
+	// Connection to the postgres database used for persistence.
+	Db *pgxpool.Pool
+	// For each partition, store the id of the most recent message.
+	// Used to detect unexpected seeks.
+	lastMessageIdByPartition map[int32]pulsar.MessageID
+	// Pulsar consumer on which to receive messages.
+	consumer pulsar.Consumer
+	// Optional logger.
+	// If not provided, the default logrus logger is used.
+	Logger *logrus.Entry
+}
+
+// Batch of changes to be written to postgres.
+type Batch struct {
+	// Time at which this batch was created.
+	CreatedAt time.Time
+	// Ids of messages processed to create this batch.
+	// Note that a batch may contain several message ids but not changes to be written to postgres.
+	// Since only a subset of messages result in changes to postgres.
+	MessageIds []pulsar.MessageID
+	// New jobs to be inserted.
+	// Should always be inserted first.
+	Jobs []Job
+	// New job runs to be inserted.
+	// Should be inserted after jobs.
+	Runs []Run
+	// Reprioritisations.
+	// When writing to postgres, the priority that was last written to Pulsar for each job wins out.
+	// For ReprioritiseJobSet, the priority
+	Reprioritisations []*ReprioritisationBatch
+	// Set of job sets to be canceled.
+	// The map is used as a set, i.e., the value of the bool doesn't matter.
+	JobSetsCancelled map[string]bool
+	// Set of jobs to be canceled.
+	// The map is used as a set, i.e., the value of the bool doesn't matter.
+	JobsCancelled map[uuid.UUID]bool
+	// Set of jobs that have succeeded.
+	JobsSucceeded map[uuid.UUID]bool
+	// Any job error messages received.
+	JobErrors []JobError
+	// Map from run id to a struct describing the set of physical resources assigned to that run.
+	JobRunAssignments map[uuid.UUID]*JobRunAssignment
+	// Ids of job runs that have started running.
+	JobRunsRunning map[uuid.UUID]bool
+	// Ids of job runs that have succeeded.
+	JobRunsSucceeded map[uuid.UUID]bool
+	// Any job run error messages received.
+	JobRunErrors []JobRunError
+}
+
+func NewBatch() *Batch {
+	return &Batch{
+		CreatedAt:         time.Now(),
+		JobSetsCancelled:  make(map[string]bool),
+		JobsCancelled:     make(map[uuid.UUID]bool),
+		JobsSucceeded:     make(map[uuid.UUID]bool),
+		JobRunAssignments: make(map[uuid.UUID]*JobRunAssignment),
+		JobRunsRunning:    make(map[uuid.UUID]bool),
+		JobRunsSucceeded:  make(map[uuid.UUID]bool),
+	}
+}
+
+// Batch of reprioritisations.
+// PrioritiesByJobSet should always be applied before PrioritiesByJob.
+// ReprioritisationBatch
+// To ensure changes are applied in the correct order.
+type ReprioritisationBatch struct {
```

Comment:
I'd be tempted to rename this `Reprioritisations` or something.  From the names I initially assumed this would be a `Batch` implementation